codex-rs/network-proxy/src/http_proxy.rs

```diff
@@ -0,0 +1,450 @@
+use crate::config::NetworkMode;
+use crate::mitm;
+use crate::policy::normalize_host;
+use crate::state::AppState;
+use crate::state::BlockedRequest;
+use anyhow::Context;
+use anyhow::Result;
+use anyhow::anyhow;
+use rama::Context as RamaContext;
+use rama::Layer;
+use rama::Service;
+use rama::http::Body;
+use rama::http::Request;
+use rama::http::Response;
+use rama::http::StatusCode;
+use rama::http::client::EasyHttpWebClient;
+use rama::http::layer::remove_header::RemoveRequestHeaderLayer;
+use rama::http::layer::remove_header::RemoveResponseHeaderLayer;
+use rama::http::layer::upgrade::UpgradeLayer;
+use rama::http::layer::upgrade::Upgraded;
+use rama::http::matcher::MethodMatcher;
+use rama::http::server::HttpServer;
+use rama::net::http::RequestContext;
+use rama::net::proxy::ProxyTarget;
+use rama::net::stream::SocketInfo;
+use rama::service::service_fn;
+use rama::tcp::client::service::Forwarder;
+use rama::tcp::server::TcpListener;
+use serde_json::json;
+use std::convert::Infallible;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tracing::error;
+use tracing::info;
+use tracing::warn;
+
+type ContextState = Arc<AppState>;
+type ProxyContext = RamaContext<ContextState>;
+
+pub async fn run_http_proxy(state: Arc<AppState>, addr: SocketAddr) -> Result<()> {
+    let listener = TcpListener::build_with_state(state)
+        .bind(addr)
+        .await
+        .map_err(|err| anyhow!("bind HTTP proxy: {err}"))?;
```

Comment:
Prefer anyhow's `.context()` or `.with_context()` over `.map_err()` so we get proper chaining.